FIX/TST: fix 'atef check' and add it to the test suite

klauer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
Tests and fix for command-line entrypoints.

## Motivation and Context
* `atef check` has been broken
* Closes #29
* Restructures how PV/Device -> Comparison works by adding another layer called `PreparedComparison`
* This allows the caller to have more control over which comparisons get run and when

## How Has This Been Tested?
* Test suite
* Also, local testing with `atef check`

## Questions
* Is there a better way to do the yaml stuff that I'm missing?
* Is there a more obvious API that could be used?

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/5139267/164332592-de0fe5ac-3805-4cac-a878-aeb408f8cf79.png)

klauer commented:
This isn't quite there yet, but I'm open to suggestions API-wise on how to proceed. 

I think the `PreparedComparison` (however awkwardly named) is an improvement to the single option which was supported previously (`check_pvs` or `check_device`; these remain in this PR to avoid rewriting the test suite. They are refactored around the new code path.).

As noted in the PR text above, I'm hoping that yielding everything that is to be run could offer flexibility for various implementations (command-line or GUI) to decide what checks are to be run and when to execute them.

One downside I can see, so far, is that using `PreparedComparison.from_config()` to get a bunch of `PreparedComparison` instances removes some of the hierarchy inherent in the dataclass structure: it's like taking the outer product of all possible comparisons. It'd be up to the caller to try and reassemble that information or keep track of it.

Is there enough here to make sense of what this does, or would this benefit from more of a description (or a meeting)?

ZLLentz commented:
I'm going to look through this and give a review best I can. If I fail we can have a meeting.
The changes here make me realize that I should be at least putting out a "make the cfg gui Window" test since I was importing a lot of stuff from `check`. I'll fix those imports and add at least basic tests in a different PR.

ZLLentz commented:
OK, I do like this. I'll need to look a second time for line-by-line suggestion slotting, but nothing stuck out as bad/commentable on the first pass (high-level pass, focused on understanding). I imagine that the import failures were related to the dichotomy between PV checks and device checks and that led you to realize that bigger changes were needed?

It seems like you haven't "changed" any of the `atef.check` or `ConfigurationFile` internals, you've refactored a bit and added better interfacing for loading all the tests and for disambiguating a test load failure from other failures. Is this a fair summary?

I like how flexible the output will be- I'm not forced to run any particular test (or, even, any tests at all) just because it's in the file. The config gui could have a tab that uses this to summarize all the configured tests, for example.

Small food for thought: this interface path does require us to (try to) instantiate all of the devices/signals, in order. So, much like other applications, if bad/slow devices are included in the chain we can potentially have a slow load, which is consequential if there is only a subset of the tests to run. We should fix our devices, of course, rather than worrying about this in every application.

> One downside I can see, so far, is that using `PreparedComparison.from_config()` to get a bunch of `PreparedComparison` instances removes some of the hierarchy inherent in the dataclass structure: it's like taking the outer product of all possible comparisons. It'd be up to the caller to try and reassemble that information or keep track of it.

The only real consequence of this that I see is that it's not possible to reconstruct the exact file config from a list of the prepared comparisons. That is, more than one file could result in the same test suite. But this isn't too surprising because the file contains a lot of metadata. It is still possible to identify which branch of the config file a particular `PreparedComparison` came from by comparing the `comparison` attribute with elements from the config file, and most callers here probably have access to the original data structure.

klauer commented:
> OK, I do like this. I'll need to look a second time for line-by-line suggestion slotting, but nothing stuck out as bad/commentable on the first pass (high-level pass, focused on understanding). 

Great! That's a good sign.

> I imagine that the import failures were related to the dichotomy between PV checks and device checks and that led you to realize that bigger changes were needed?

It was something like:
1. Wow, I didn't update `atef.bin.check` and refactored all of the other internals, oops
2. Oh, this API isn't very nice, and I can't get detailed source information easily from a `Result` (verbose mode shows "Success: none" for successful checks?!)
3. Hmm, what if the GUI wanted to use this?
4. Refactor, refactor, refactor...

> It seems like you haven't "changed" any of the atef.check or ConfigurationFile internals, you've refactored a bit and added better interfacing for loading all the tests and for disambiguating a test load failure from other failures. Is this a fair summary?

That's a fair summary, indeed. The key API additions are there to allow for more unified interactions with the combinations of `PV + Comparison` or `Device + Attribute + Comparison`.

> Small food for thought: this interface path does require us to (try to) instantiate all of the devices/signals, in order. So, much like other applications, if bad/slow devices are included in the chain we can potentially have a slow load, which is consequential if there is only a subset of the tests to run. We should fix our devices, of course, rather than worrying about this in every application.

Agreed on "optimize the common problem". 

It does make me wonder about how subprocess handling could work here as our information should be easily serialized and sent back to the main process. Our verbiage, so to speak, is significantly less than the average bluesky session. "Make a device, run this comparison against its signal and gie me the result"

ZLLentz commented:
> I think the key part that you nailed is most callers here probably have access to the original data structure.. That probably will be enough for most uses. If not, we can revisit down the line, I suppose.

The use case here is probably something like "Where did I define my test?", right? It should be possible to add helper functions to `ConfigurationFile` for solving this, or maybe even the context information could be stored on the `Comparison` instance itself at runtime.

klauer commented:
I think I could shuffle this code around for a lot longer and still not be happy with it.
So I'd say if all of this looks good, let's merge, move on, and revisit as appropriate.

ZLLentz commented:
I'll do a more complete review then, to the extent I'm able given that I'm not 100% familiar with all the code and that refactors are hard to review

ZLLentz commented on MANIFEST.in:
```diff
@@ -12,7 +12,7 @@ recursive-include docs *.rst conf.py Makefile make.bat
 include atef/_version.py
 
-include atef/tests/*.json
+recursive-include atef/tests *.json *.yml
```
I didn't know about this, neat trick

ZLLentz commented on atef/bin/__init__.py:
```diff
@@ -1,3 +1,3 @@
-from .main import main
+from . import main
```
This is a functionally different import than before, right? (`atef.bin.main` module vs the contained `main` function). Is this just to remove the ambiguity of statements like `from atef.bin import main`? Does this import need to be here at all?

klauer replied:
It is different - and part of a weird bug (or lack of understanding on my part).

Importing `main.main` as `main` effectively shadowed the submodule, and I wasn't able to access `atef.bin.main.something_else`. I switched it up to this instead.

I think it's not a necessary import. My only reason for keeping it would be that it's nice to have something that tab-completes when you import a submodule.

klauer commented on atef/check.py:
```diff
@@ -780,3 +1071,28 @@ def get_signal_cache() -> _SignalCache[ophyd.EpicsSignalRO]:
+        """Helper for pyyaml to represent string enums as just strings."""
+        return dumper.represent_str(data.value)
+
+    # The ugliness of this makes me think we should use a different library
```
A few possibilities, zero or more could be right:
* I may be missing a simpler/easier option for this
* The pyyaml library just isn't very nice (and maybe we should take a look at like `ruamel.yaml` which seems popular?)
* Maybe we drop yaml entirely? I do like the sample file format better than JSON though

ZLLentz replied:
Without this weirdness, the yaml file is not reasonably human editable, right?

If ruamel.yaml gave a natively human readable/writable output I'd be in favor of switching.
I don't think it matters for this PR though.

I don't have an answer to dropping yaml or keeping it. It seems annoying to have multiple formats but it also doesn't seem like that much of a problem... I was thinking about making the GUI read both and save json by default or yaml if it sees .yml or .yaml

ZLLentz replied:
Actually maybe the inverse is better: yaml by default, and the ability to switch to json in case your config file gets large enough to have a slow load.

klauer replied:
> Without this weirdness, the yaml file is not reasonably human editable, right?

It has weird 'tags' in there indicating how to recreate the Python enum object. "Human editable with frustration" may be appropriate.

A couple details I neglected to note:
* With those enum tags, `yaml.safe_load` doesn't work as it might allow for arbitrary code execution.
* As integer values in the yaml file, apischema dutifully converts these for us on deserialization, so storage as an integer is not a problem.

> If ruamel.yaml gave a natively human readable/writable output I'd be in favor of switching. I don't think it matters for this PR though.

I'd like to investigate it to see how much better it actually is. Agreed on being out of scope here.

> Actually maybe the inverse is better: yaml by default, and the ability to switch to json in case your config file gets large enough to have a slow load.

Yeah, that could be a reasonable approach.

I think we're both on the same page for "let's not make any big decisions on file formats just yet". Before we have tons of atef configuration files in the wild, though, we really should figure it out.

ZLLentz commented on atef/bin/check.py:
```diff
@@ -42,14 +39,15 @@ def build_arg_parser(argparser=None):
         "-v", "--verbose",
         action="count",
         help="Increase output verbosity",
+        default=0,
```
Out of scope for this PR: is it simple for us to define common arguments like `verbose` in a more central place? Is that what the `argparser` argument is for here?

klauer replied:
It could be moved to `atef.bin.main` pretty easily. The only annoying thing might be that the verbosity level comes before the command instead of after:
`$ atef -v check ...`

ZLLentz commented on atef/bin/check.py:
```diff
@@ -208,46 +83,137 @@ def log_results_rich(
+        severity=get_maximum_severity(severities),
+        items=items,
+        verbose=verbose,
+    )
```
A lot of the stuff in this section above I'll need to play around with manually to get a feel for/opinions on it. Should merge without considering that and follow-up later if needed.

ZLLentz commented on atef/type_hints.py:
```diff
@@ -1,4 +1,6 @@
+import pathlib
 from typing import Union
 
+AnyPath = Union[str, pathlib.Path]
```
All of these type hint shortcuts should really be in the core python typing module, in my opinion. They are broadly useful.

klauer replied:
Totally agreed...

ZLLentz commented on atef/check.py:
```diff
@@ -631,6 +648,285 @@ def get_by_tag(self, *tags: str) -> Generator[Configuration, None, None]:
                 yield config
 
+    @classmethod
+    def from_json(cls, filename: AnyPath) -> ConfigurationFile:
```
Good call on these, I'll switch the gui to use these methods instead of re-implementing them

ZLLentz commented on atef/check.py:
```diff
@@ -631,6 +648,285 @@ def get_by_tag(self, *tags: str) -> Generator[Configuration, None, None]:
+        *,
+        client: Optional[happi.Client] = None,
+        cache: Optional[Mapping[str, ophyd.Signal]] = None,
+    ) -> Generator[Union[PreparedComparison, PreparedComparisonException], None, None]:
```
I'm a big fan of how this class works

ZLLentz reviewed (approved):
I left a lot of comments and some questions, I don't actually remember what they were. I think this is good to merge.

ZLLentz commented:
Ok I think we merge this, I'll do a follow-up to fix the gui's import paths and add a really basic "can we make a window" unit test

klauer commented:
Here we go...
